### PR TITLE
First Version Using IBGE Censo 2010 Analytical Data And Custom Gender Detector As Fallback

### PR DESCRIPTION
**Pull request by democlitos**
The comments on the code are yard compatible. Run ```yard``` to generate the related documentation.

### CODE REVIEW

File: lib/name_gender_classifier.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
+require 'iconv'
+
+# Gender detector for first names.
+module NameGenderClassifier
+  # Return the gender(s) (probabilistically) for the informed name(s). The result type will vary depending
```
Comment (pacarvalho) [inline lib/name_gender_classifier.rb:5]: probabilistically -> probability

File: lib/name_gender_classifier.rb [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  # on the parameter type:
+  #
+  # [String, Symbol] the gender (String) is returned.
+  # [Array<String>] an array (Array<String>) with the genders is returned.
+  # [Array<Object>] an array (Array<Object>) with the same objects and the newly assigned genders is returned.
+  #
+  # @param arg [String, Symbol, Array<String>, Array<Symbol>, Array<Object>] argument holding first
+  #   name(s) information(s).
+  # @paran options [Hash] first_name_attribute: name of the method that returns the first name,
```
Comment (pacarvalho) [inline lib/name_gender_classifier.rb:14]: @param typo

File: lib/name_gender_classifier.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  #                       gender_attribute: name of the method which will receive the gender assignment.
+  #
+  # @return [String, Array<String>, Array<Object>] the gender classification for the passed first names
+  def self.classify(arg, options = {})
+    case arg
+    when String, Symbol
+      most_probable_gender(arg)
+    when Array
+      if arg[0].is_a?(String) || arg[0].is_a?(Symbol)
```
Comment (pacarvalho) [inline lib/name_gender_classifier.rb:23]: Add a comment above line 23 that says:

# Assumes that all elements within the array are of the same type as the first.

File: lib/name_gender_classifier.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  #
+  # @param objects [Array<Object>] see {NameGenderClassifier.classify}
+  # @param options see {NameGenderClassifier.classify}
+  #
+  # @return [Array<Object>] the objects with the assigned genders
+  def self.classify_objects(objects, options = {})
+    gender_attribute = options.fetch(:gender_attribute, 'gender')
+    first_name_attribute = options.fetch(:first_name_attribute, nil)
+    gender_attribute_assignment = "#{gender_attribute}="
+    if first_name_attribute.nil?
```
Comment (pacarvalho) [inline lib/name_gender_classifier.rb:60]: I feel like there may be a way to implement the fallbacks to `first_name_attributes` without the if statement. 

What if you do: 

first_name_attribute = options.fetch(:first_name_attribute, nil) || (:first_name if defined?(objects[0].first_name)) || (:name if defined?(objects[0].name))

File: lib/name_gender_classifier.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        object.public_send(gender_attribute_assignment, most_probable_gender(name, db))
+      end
+    end
+
+    objects
+  end
+
+  # Remove whitespaces, secondary names, accents, digits and transform to string and lower case.
+  def self.remove_unwanted_chars(name)
+    Iconv.iconv('ascii//translit//ignore', 'utf-8', name.to_s.strip.split(' ')[0].downcase)[0].gsub(/\W+/, '')
```
Comment (pacarvalho) [inline lib/name_gender_classifier.rb:86]: Why do you need Iconv here? Without it, we would only have 1 dependency.

File: name_gender_classifier.gemspec [hunk 6]
```diff
@@ -0,0 +1,17 @@
+Gem::Specification.new do |s|
+  s.name = 'name_gender_classifier'
+  s.version     = '0.0.1'
+  s.summary     = 'Detecção de gênero pt-br.'
```
Comment (pacarvalho) [inline name_gender_classifier.gemspec:4]: Gender detection for brazilian first names.

File: name_gender_classifier.gemspec [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  s.description = 'Utilizando a base de dados do censo IBGE 2010, classificamos primeiros nomes como :male ou :female'
```
Comment (pacarvalho) [inline name_gender_classifier.gemspec:5]: Translate

File: name_gender_classifier.gemspec [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+  s.authors     = ['Avantsoft']
+  s.email       = 'admin@avantsoft.com.br'
```
Comment (pacarvalho) [inline name_gender_classifier.gemspec:7]: hello@

File: lib/name_gender_classifier.rb (hunk 5 above)
Comment (democlitos) [reply lib/name_gender_classifier.rb:86]: Just to transliterate words. For example, transforming 'Corção' into 'Corcao'.

**Review by pacarvalho — APPROVED** [review]

**Pull request merged**